suite/tests/linux/rseq.c

```diff
@@ -271,6 +269,71 @@ test_rseq_branches(void)
     assert(completions == 1 && restarts > 0 && sigill_count == 1);
 }
 
+/* Tests that DR handles a signal inside its native rseq copy.
+ * Any synchronous signal is going to pretty much never happen for real, since
+ * it would happen on the instrumentation execution and never make it to the
+ * native run, but an asynchronous signal could arrive.
+ * It's complicated to set up an asynchronous signal at the right spot, so
+ * we cheat and take advantage of DR not restoring XMM state to have different
+ * behavior in the two DR executions of the rseq code.
+ */
+static void
+test_rseq_native_fault(void)
+{
+    int restarts = 0;
+    __asm__ __volatile__(
+        /* clang-format off */ /* (avoid indenting next few lines) */
+        RSEQ_ADD_TABLE_ENTRY(fault, 2f, 3f, 4f)
+        /* clang-format on */
+
+        "6:\n\t"
+        /* Store the entry into the ptr. */
+        "leaq rseq_cs_fault(%%rip), %%rax\n\t"
+        "movq %%rax, %[rseq_tls]\n\t"
+        "pxor %%xmm0, %%xmm0\n\t"
+        "mov $1,%%rcx\n\t"
+        "movq %%rcx, %%xmm1\n\t"
+
+        /* Restartable sequence. */
+        "2:\n\t"
+        /* Increase xmm0 every time.  DR (currently) won't restore xmm inputs
```

Comment:
Clever! Do we want to document this assumption somewhere?

Reply:
We do document the GPR limitation in the public docs in the section "Restartable Sequence Instrumentation Constraints" in bt.dox, and point at that list from the wiki doc https://github.com/DynamoRIO/dynamorio/wiki/Restartable-Sequences.  Looking at it: it is slightly stale since it talks about read-before-written.  I will update it separately.